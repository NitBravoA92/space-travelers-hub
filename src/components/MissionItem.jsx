import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';
import { addReservedMission, LeaveMission } from '../redux/missions/missionsSlice';

const MissionItem = ({ mission }) => {
  const dispatch = useDispatch();
  const handleClickAddReservedMission = () => {
    dispatch(addReservedMission(mission.mission_id));
  };

  const handleClickLeaveMission = () => {
    dispatch(LeaveMission(mission.mission_id));
  };

  return (
    <tr>
      <td className="mission-name">{ mission.mission_name }</td>
      <td className="mission-description">{ mission.description }</td>
      <td className="mission-status">
        {
          mission.reserved ? (
            <span className="badge badge-blue active-member">Active member</span>
          )
            : (
              <span className="badge badge-gray">Not a member</span>
            )
        }
      </td>
      <td className="mission-join-button">
        {mission.reserved ? (
          <button type="button" className="btn btn-pink" onClick={handleClickLeaveMission}>Leave Mission</button>
        )
          : (
            <button type="button" className="btn btn-gray" onClick={handleClickAddReservedMission}>Join Mission</button>
          )}
      </td>
    </tr>
  );
};

MissionItem.propTypes = {
  mission: PropTypes.shape({
    mission_id: PropTypes.string.isRequired,
    mission_name: PropTypes.string.isRequired,
    description: PropTypes.string.isRequired,
    reserved: PropTypes.bool,
  }).isRequired,
};

export default MissionItem;
