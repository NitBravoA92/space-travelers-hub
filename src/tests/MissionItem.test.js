import '@testing-library/jest-dom/extend-expect';
import { screen } from '@testing-library/react';
import MissionItem from '../components/MissionItem';
import renderWithProviders from '../helpers/helper-for-test';

describe('The MissionItem component', () => {
  const mockData = {
    mission_id: '9D1B7E0',
    mission_name: 'Telstar 19V',
    description:
      'Thaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited.',
  };

  test('renders correctly into the DOM', () => {
    const { tree } = renderWithProviders(
      <table>
        <tbody>
          <MissionItem mission={mockData} />
        </tbody>
      </table>,
    );
    expect(tree).toMatchSnapshot();
  });

  test("renders correctly into the DOM showing the text: 'Telstar 19V' as the mission title", () => {
    renderWithProviders(
      <table>
        <tbody>
          <MissionItem mission={mockData} />
        </tbody>
      </table>,
    );

    const missionTitle = screen.getByText(/Telstar 19V/i);

    expect(missionTitle).toBeInTheDocument();
  });

  test("renders the text 'Not a member' when the user doesn't have the mission joined", () => {
    renderWithProviders(
      <table>
        <tbody>
          <MissionItem mission={mockData} />
        </tbody>
      </table>,
    );

    const grayBadge = screen.getByText(/Not a member/i);
    const badgeClasses = grayBadge.className;

    expect(grayBadge).toBeInTheDocument();
    expect(badgeClasses).toMatch(/badge badge-gray/i);
  });

  test("renders a button with the text 'Join Mission'", () => {
    renderWithProviders(
      <table>
        <tbody>
          <MissionItem mission={mockData} />
        </tbody>
      </table>,
    );

    const joinButton = screen.getByRole('button', {
      name: /Join Mission/i,
    });
    const buttonClasses = joinButton.className;

    expect(joinButton).toBeInTheDocument();
    expect(buttonClasses).toMatch(/btn btn-gray/i);
  });
});
