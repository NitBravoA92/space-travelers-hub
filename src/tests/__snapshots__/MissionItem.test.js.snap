// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The MissionItem component renders correctly into the DOM 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <table>
        <tbody>
          <tr>
            <td
              class="mission-name"
            >
              Telstar 19V
            </td>
            <td
              class="mission-description"
            >
              Thaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited.
            </td>
            <td
              class="mission-status"
            >
              <span
                class="badge badge-gray"
              >
                Not a member
              </span>
            </td>
            <td
              class="mission-join-button"
            >
              <button
                class="btn btn-gray"
                type="button"
              >
                Join Mission
              </button>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </body>,
  "container": <div>
    <table>
      <tbody>
        <tr>
          <td
            class="mission-name"
          >
            Telstar 19V
          </td>
          <td
            class="mission-description"
          >
            Thaicom is the name of a series of communications satellites operated from Thailand, and also the name of Thaicom Public Company Limited.
          </td>
          <td
            class="mission-status"
          >
            <span
              class="badge badge-gray"
            >
              Not a member
            </span>
          </td>
          <td
            class="mission-join-button"
          >
            <button
              class="btn btn-gray"
              type="button"
            >
              Join Mission
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
